source/detail/parsers.cpp

```diff
@@ -0,0 +1,766 @@
+        || std::is_same<double, typename std::remove_cv<T>::type>::value
+        || std::is_same<long double, typename std::remove_cv<T>::type>::value,
+        "only float, double and long double can be parsed by the C parsers");
+
+    auto &errno_ref = errno; // Nonzero cost, pay it once; note: since c++11, errno is thread-local and thus safe
+    errno_ref = 0;
+
+    bool ok = string != nullptr;
+    assert(ok); // accepting nullptr strings hides bugs
+
+    if (ok)
+    {
+        // Note: caching the decimal separator used by the parsing functions can cause thread-safety issues in multi-threaded programs. On the other hand,
+        // this would also mean that the rest of this function's logic would be wrong in such cases. Since this isn't something we can fix in a thread-safe way,
+        // because we cannot control how users of our library call locale-dependent functions, this is a risk we have to take.
+        const char parser_decimal_separator = localeconv()->decimal_point[0];
+        const char decimal_separator = std::use_facet<std::numpunct<char>>(loc).decimal_point();
+        bool tried_dot = false;
+        bool tried_comma = false;
+
+        // Step 1: try to parse the number as it is.
+        char *internal_end = nullptr;
+        T parsed = func(string, &internal_end);
+
+        if (parser_decimal_separator == '.')
+        {
+            tried_dot = true;
+        }
+        else if (parser_decimal_separator == ',')
+        {
+            tried_comma = true;
+        }
+
+        // Step 2: if the string was only partially parsed, try to replace the decimal separator by the one from the provided locale.
+        if (*internal_end == decimal_separator && parser_decimal_separator != decimal_separator)
+        {
+            std::string replaced(string);
+            std::replace(replaced.begin(), replaced.end(), decimal_separator, parser_decimal_separator);
+            parsed = func(replaced.c_str(), &internal_end);
+
+            if (decimal_separator == '.')
+            {
+                assert(!tried_dot);
+                tried_dot = true;
+            }
+            else if (decimal_separator == ',')
+            {
+                assert(!tried_comma);
+                tried_comma = true;
+            }
+
+            // The end has to point to the end of the original string, not the replaced one.
+            // The great C Standard Library is not const correct, so we can't be either... ugh. Note that you are NOT allowed to modify
+            // the pointer to the end (thus modifying the string) without resulting in undefined behaviour!
+            internal_end = const_cast<char *>(string + (internal_end - replaced.c_str()));
+        }
+
+        // Step 3: if the string was only partially parsed even when using the provided locale,
+        // maybe the locale has the wrong decimal separator and . was the decimal separator?
+        if (*internal_end == '.' && !tried_dot)
```

Comment:
The thing is, writing a perfect parsing function is probably impossible, so in this function I'm trying to find out whether we may have parsed the number incorrectly. A number might contain thousands separators and decimal separators, and deciding whether a number has been parsed 100% correctly is rather hard. This is why I'm trying to parse the number multiple times, but in most cases it should only be parsed 1-2 times at most (since I'm checking whether we have already tried to use certain separators for parsing).

However, since we already discussed above that all serialized numbers should not contain any thousands separators and only `.` as the decimal separator, I can change this logic - see above.